File: Scala/src/main/scala/AlgorithmicHeights/INS.scala
package AlgorithmicHeights

object INS {
  import scala.annotation.tailrec

  private def convertToIntArray(line: String): Array[Int] = line.split(" ").map(_.toInt)

  def calcNrSwapsInInsertionSort(array: Array[Int], n: Int): Int = {
    @tailrec
    def moveElemToItsPlace(nrSwaps: Int, k: Int): Int = {
      if (k == 0 || array(k) >= array(k - 1)) nrSwaps
      else {
        swapElems(k, k - 1)
        moveElemToItsPlace(nrSwaps + 1, k - 1)
      }
    }

    def swapElems(i: Int, j: Int): Unit = {
      val elem: Int = array(i)
      array(i) = array(j)
      array(j) = elem
    }

    (1 until n).foldLeft(0)((acc, ix) => moveElemToItsPlace(acc, ix))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val array: Array[Int] = convertToIntArray(reader.next())
    val result: Int = calcNrSwapsInInsertionSort(array, n)
    println(result)
  }
}

File: OCaml/src/algorithmic_heights/ins.ml
let convert_to_ints (line : string) : int array =
    Array.of_list @@ List.map int_of_string Str.(line |> split (regexp " "))


let calc_nr_swaps_in_insertion_sort (array : int array) (n : int) : int =
    let swap i j =
        let elem = array.(i) in
        array.(i) <- array.(j);
        array.(j) <- elem in
    let rec move_elem_to_its_place nr_swaps k =
        if k = 0 || array.(k) >= array.(k - 1) then nr_swaps
        else
            (swap k (k - 1);
            move_elem_to_its_place (nr_swaps + 1) (k - 1)) in
    Seq.(fold_left move_elem_to_its_place 0 @@ init (n - 1) (fun ix -> ix + 1))


let () =
    let n = read_int () in
    let array = convert_to_ints (read_line ()) in
    let result = calc_nr_swaps_in_insertion_sort array n in
    print_int result; print_newline ()
